cache/cache_test.go

```diff
@@ -0,0 +1,369 @@
+					expectedCurSize++
+				} else {
+					require.NotNil(t, actualResult)
+					
+					// Here, op.expectedResult represents the index of the removed node in tc.cacheOps
+					require.Equal(t, testNodes[int(op.expectedResult)], actualResult)
+				}
+				require.Equal(t, expectedCurSize, cache.Len())
+			}
+
+			validateCacheContentsAfterTest(t, tc, cache)
+		})
+	}
+}
+
+func Test_Cache_Remove(t *testing.T) {
+	testcases := map[string]testcase{
+		"remove non-existent key, cache limit 0 - nil returned": {
+			cacheLimit: 0,
+			cacheOps: []cacheOp{
+				{
+					testNodexIdx: 0,
+					expectedResult: noneRemoved,
+				},
+			},
+		},
+		"remove non-existent key, cache limit 1 - nil returned": {
+			setup: func(c cache.Cache) {
+				require.Nil(t, c.Add(testNodes[1]))
+				require.Equal(t, 1, c.Len())
+			},
+			cacheLimit: 1,
+			cacheOps: []cacheOp{
+				{
+					testNodexIdx: 0,
+					expectedResult: noneRemoved,
+				},
+			},
+			expectedNodeIndexes: []int{1},
+		},
+		"remove existent key, cache limit 1 - removed": {
+			setup: func(c cache.Cache) {
+				require.Nil(t, c.Add(testNodes[0]))
+				require.Equal(t, 1, c.Len())
+			},
+			cacheLimit: 1,
+			cacheOps: []cacheOp{
+				{
+					testNodexIdx: 0,
+					expectedResult: 0,
+				},
+			},
+		},
+		"remove twice, cache limit 1 - removed first time, then nil": {
+			setup: func(c cache.Cache) {
+				require.Nil(t, c.Add(testNodes[0]))
+				require.Equal(t, 1, c.Len())
+			},
+			cacheLimit: 1,
+			cacheOps: []cacheOp{
+				{
+					testNodexIdx: 0,
+					expectedResult: 0,
+				},
+				{
+					testNodexIdx: 0,
+					expectedResult: noneRemoved,
+				},
+			},
+		},
+		"remove all, cache limit 3": {
+			setup: func(c cache.Cache) {
+				require.Nil(t, c.Add(testNodes[0]))
+				require.Nil(t, c.Add(testNodes[1]))
+				require.Nil(t, c.Add(testNodes[2]))
+				require.Equal(t, 3, c.Len())
+			},
+			cacheLimit: 3,
+			cacheOps: []cacheOp{
+				{
+					testNodexIdx: 2,
+					expectedResult: 2,
+				},
+				{
+					testNodexIdx: 0,
+					expectedResult: 0,
+				},
+				{
+					testNodexIdx: 1,
+					expectedResult: 1,
+				},
+			},
+		},
+	}
+
+	for name, tc := range testcases {
+		t.Run(name, func(t *testing.T) {
+			cache := cache.New(tc.cacheLimit)
+
+			if tc.setup != nil {
+				tc.setup(cache)
+			}
+
+			expectedCurSize := cache.Len()
+
+			for _, op := range tc.cacheOps {
+
+				actualResult := cache.Remove(testNodes[op.testNodexIdx].GetKey())
+
+				expectedResult := op.expectedResult
+
+				if expectedResult == noneRemoved {
+					require.Nil(t, actualResult)
+				} else {
+					expectedCurSize--
+					require.NotNil(t, actualResult)
+					
+					// Here, op.expectedResult represents the index of the removed node in tc.cacheOps
+					require.Equal(t, testNodes[int(op.expectedResult)], actualResult)
+				}
+				require.Equal(t, expectedCurSize, cache.Len())
+			}
+
+			validateCacheContentsAfterTest(t, tc, cache)
+		})
+	}
+}
+
+func BenchmarkAdd(b *testing.B) {
```

Comment:
I've typically seen benchmarks separated out into a dedicated `*_bench_test.go` file so they don't clobber with the unit tests. WDYT?

Reply:
Good idea. Although this is not a convention in this repo, I think it's good to slowly start doing that. Moved cache benchmarks into a separate file `cache_bench_test.go`